internal/reliable_topic.go

```diff
@@ -0,0 +1,284 @@
+		registrationID, nil)
+}
+
+func (r *ReliableTopicProxy) Publish(message interface{}) (err error) {
+	messageData, err := r.validateAndSerialize(message)
+	if err != nil {
+		return err
+	}
+	reliableMsg := reliabletopic.NewMessage(messageData, nil)
+	switch r.topicOverLoadPolicy {
+	case core.TopicOverLoadPolicyError:
+		err = r.addOrFail(reliableMsg)
+	case core.TopicOverLoadPolicyBlock:
+		err = r.addWithBackoff(reliableMsg)
+	case core.TopicOverLoadPolicyDiscardNewest:
+		_, err = r.ringBuffer.Add(reliableMsg, core.OverflowPolicyFail)
+	case core.TopicOverLoadPolicyDiscardOldest:
+		err = r.addOrOverwrite(reliableMsg)
+	}
+	return err
+}
+
+func (r *ReliableTopicProxy) addOrFail(message *reliabletopic.Message) (err error) {
+	seqID, err := r.ringBuffer.Add(message, core.OverflowPolicyFail)
+	if err != nil {
+		return err
+	}
+	if seqID == -1 {
+		// TODO :: add message to error string
+		errorMsg := "failed to publish message to topic: " + r.name
+		return core.NewHazelcastTopicOverflowError(errorMsg, nil)
+	}
+	return
+}
+
+func (r *ReliableTopicProxy) addWithBackoff(message *reliabletopic.Message) (err error) {
+	sleepTime := initialBackoff
+	for {
+		result, err := r.ringBuffer.Add(message, core.OverflowPolicyFail)
+		if err != nil {
+			return err
+		}
+		if result != -1 {
+			return nil
+		}
+		time.Sleep(sleepTime)
+		sleepTime *= 2
+		if sleepTime > maxBackoff {
+			sleepTime = maxBackoff
+		}
+	}
+}
+
+func (r *ReliableTopicProxy) addOrOverwrite(message *reliabletopic.Message) (err error) {
+	_, err = r.ringBuffer.Add(message, core.OverflowPolicyOverwrite)
+	return
+}
+func (r *ReliableTopicProxy) Ringbuffer() core.Ringbuffer {
+	return r.ringBuffer
+}
+
+type messageProcessor struct {
+	id        string
+	sequence  int64
+	cancelled atomic.Value
+	listener  core.ReliableMessageListener
+	proxy     *ReliableTopicProxy
+}
+
+func newMessageProcessor(id string, listener core.ReliableMessageListener, proxy *ReliableTopicProxy) *messageProcessor {
+	msgProcessor := &messageProcessor{
+		id:       id,
+		listener: listener,
+		proxy:    proxy,
+	}
+	msgProcessor.cancelled.Store(false)
+	initialSeq := listener.RetrieveInitialSequence()
+	if initialSeq == -1 {
+		// set initial seq as tail sequence + 1 so that we wont be reading the items that were added before the listener
+		// is added.
+		tailSeq, _ := msgProcessor.proxy.ringBuffer.TailSequence()
+		initialSeq = tailSeq + 1
+	}
+	msgProcessor.sequence = initialSeq
+	return msgProcessor
+}
+
+func (m *messageProcessor) next() {
+	if m.cancelled.Load() == true {
+		return
+	}
+	readResults, err := m.proxy.ringBuffer.ReadMany(m.sequence, 1,
+		m.proxy.config.ReadBatchSize(), nil)
+	if err == nil {
+		m.onResponse(readResults)
+	} else {
+		m.onFailure(err)
+	}
+}
+
+func (m *messageProcessor) onFailure(err error) {
+	if m.cancelled.Load() == true {
+		return
+	}
+	if strings.Contains(err.Error(), "com.hazelcast.ringbuffer.StaleSequenceException") {
```

Comment:
string comparison? isn't there some kind of error code to compare?

Reply:
We can get that information from the stack trace but currently there is no way of accessing the parts of stack trace and the whole stack trace is a single string. I will create an issue for this so that we can get individual parts of stack trace.